samza-api/src/main/java/org/apache/samza/operators/windows/internal/WindowInternal.java

```diff
@@ -55,28 +55,32 @@
   /*
    * The function that extracts the key from a {@link MessageEnvelope}
    */
-  private final Function<M, WK> keyExtractor;
+  private final MapFunction<M, WK> keyExtractor;
 
   /*
    * The function that extracts the event time from a {@link MessageEnvelope}
    */
-  private final Function<M, Long> eventTimeExtractor;
+  private final MapFunction<M, Long> eventTimeExtractor;
 
   /**
    * The type of this window. Tumbling and Session windows are supported for now.
    */
   private final WindowType windowType;
 
-  private final Serde<WK> keySerde;
-  private final Serde<WV> windowValSerde;
-  private final Serde<M> msgSerde;
-
   private Trigger<M> earlyTrigger;
   private Trigger<M> lateTrigger;
   private AccumulationMode mode;
 
-  public WindowInternal(Trigger<M> defaultTrigger, Supplier<WV> initializer, FoldLeftFunction<M, WV> foldLeftFunction,
-      Function<M, WK> keyExtractor, Function<M, Long> eventTimeExtractor, WindowType windowType, Serde<WK> keySerde,
+  /**
+   * The following serdes should only be used to generate configs for store config.
+   * No need to create per-task copy of those {@link Serde} objects
```

Comment:
Maybe (everywhere):
"The following Serde fields are serialized by the ExecutionPlanner when generating the store configs, and deserialized once during startup in SamzaContainer. They don't need to be deserialized here on a per-task basis."